test/SILOptimizer/pound_assert.swift

```diff
@@ -497,3 +497,36 @@ func stringInitNonEmptyFlowSensitive() -> ContainsString {
 func invokeStringInitNonEmptyFlowSensitive() {
   #assert(stringInitNonEmptyFlowSensitive().x == 1)
 }
+
+//===----------------------------------------------------------------------===//
+// Enums and optionals.
+//===----------------------------------------------------------------------===//
+func isNil(_ x: Int?) -> Bool {
+  return x == nil
+}
+
+#assert(isNil(nil))
+#assert(!isNil(3))
+
+public enum Pet {
+  case bird
+  case cat(Int)
+  case dog(Int, Int)
+  case fish
+}
+
+public func weighPet(pet: Pet) -> Int {
+  switch pet {
+  case .bird: return 3
+  case let .cat(weight): return weight
+  case let .dog(w1, w2): return w1+w2
+  default: return 1
+  }
+}
+
+#assert(weighPet(pet: .bird) == 3)
+#assert(weighPet(pet: .fish) == 1)
+#assert(weighPet(pet: .cat(2)) == 2)
+// expected-error @+1 {{assertion failed}}
+#assert(weighPet(pet: .cat(2)) == 3)
+#assert(weighPet(pet: .dog(9, 10)) == 19)
```

Comment:
I wonder whether this patch handles indirect, recursive enums, or if it needs something more? I think it would be great to add a test for indirect, recursive enums. It may help to document the behavior on them.

Reply:
Indeed! Indirect enums seem not to work -- the thing that is immediately blocking them is that the const evaluator does not do `alloc_box`. I have added some test code but commented out the `#assert` statements that actually trigger the test. I'd prefer not to implement indirect enum support myself, and there's nothing left in the tensorflow branch to copy over, so if you want it, would you or someone else take that on after this gets merged?

Also your question reminded me to add a test for address-only enums. So I did that and that seems to work.

Reply:
Sure. Thanks for adding a test. We can add the support for indirect enum when it is needed. It is enough that we document that using test cases here.